src/deploy/functions/validate.ts

```diff
@@ -35,9 +35,15 @@ export function functionsDirectoryExists(
  * @param functionNames Object containing function names as keys.
  * @throws { FirebaseError } Function names must be valid.
  */
-export function functionIdsAreValid(functions: { id: string }[]): void {
-  const validFunctionNameRegex = /^[a-zA-Z0-9_-]{1,63}$/;
-  const invalidIds = functions.filter((fn) => !validFunctionNameRegex.test(fn.id));
+export function functionIdsAreValid(functions: { id: string; platform: string }[]): void {
+  const v1FunctionName = /^[a-zA-Z][a-zA-Z0-9_-]{0,62}$/;
+  const v2FunctionName = /^[a-z][a-z-]{0,62}$/;
+  const invalidIds = functions.filter((fn) => {
+    return (
+      (fn.platform === "gcfv1" && !v1FunctionName.test(fn.id)) ||
+      (fn.platform === "gcfv2" && !v2FunctionName.test(fn.id))
+    );
+  });
   if (invalidIds.length !== 0) {
     const msg =
       `${invalidIds.join(", ")} function name(s) can only contain letters, ` +
```

Comment:
I think you need to clarify the error message for the v2 case (add a specific "v2 function names can only contain lowercase letters, numbers, and hyphens") or it will be confusing.